src/Data/Arity/Linear/Internal.hs

```diff
@@ -19,3 +62,26 @@ type family Arity b f where
           ':<>: 'ShowType b
           ':<>: 'Text "."
       )
+
+-- | The 'IsFun' type class is meant to help the type checker fill in
+-- blanks. Chances are that you can safely ignore 'IsFun' completely if it's in
+-- the type of a function you care. But read on if you are curious.
+--
+-- The type class 'IsFun' is a kind of inverse to 'FunN', it is meant to be
+-- read as @'IsFunN' a b f@ if and only if there exists `n` such that @f =
+-- 'FunN' n a b@ (`n` can be retrieved as @'Arity' b f@).
+--
+-- The reason why 'Arity' (read its documentation first) is not sufficient for
+-- our purpose, is that it can find `n` /if/ `f` is a linear function of the
+-- appropriate shape. But what if `f` is partially undetermined? Then it is
+-- likely that 'Arity' will be stuck. But we know, for instance, that if `f = a1
+-- %1 -> a2 %1 -> c` then we must have `a1 ~ a2`. The trick is that instance
+-- resolution of 'IsFun' will add unification constraints that the type checker
+-- has to solve. Look in particular at the instance @'IsFunN' a b (a' %p ->
+-- f))@: it matches liberally, so triggers on quite underdetermined `f`, but has
+-- equality constraints in its context which will help the type checker.
+class IsFunN a b f
```

Comment:
Couldn't we use something like
```haskell
class IsFunN a b f where
  type Arity

instance IsFunN a b b where
  type Arity = 'Z

instance (IsFunN a b f, a' ~ a, p ~ 'One) => IsFunN a b (a' %p -> f) where
  type Arity = 'S (Arity @a @b @f)
```
to get rid of the explicit `Arity` type family ?

Reply:
Why?

Reply:
I like the idea of having one class `FunN` to infer `f` from `n a b`, and one class `IsFunN` to infer `n` from `a b f`. Cheking `IsFunN a b f` should allow to compute the arity along the way (at least theorically).

Reply:
Did you manage to answer your own question? if so, I'm just waiting for the green tick so that I can merge.

Reply:
Yes, sorry

Reply:
> I thought that you wanted me to continue the work on this branch.
I'll pretty never be asking you to continue work on a branch which passes tests :slightly_smiling_face: . If something is mergeable as it is, then we merge it. Merging small chunks early makes everybody's life easier.